polkadot/node/subsystem-util/src/lib.rs

```diff
@@ -440,3 +442,210 @@ impl Validator {
+		block_number: BlockNumber,
+		session_index: SessionIndex,
+		para_id: ParaId,
+		validator_index: ValidatorIndex,
+	) -> ChunkIndex {
+		let shuffle = Self::get_shuffle(
+			maybe_client_features.as_ref(),
+			block_number,
+			babe_randomness,
+			n_validators,
+		);
+		self.0.insert((block_number, session_index), (shuffle, maybe_client_features));
+
+		self.query_cache_for_validator(block_number, session_index, para_id, validator_index)
+			.expect("We just inserted the entry.")
+	}
+
+	/// Return and populate the cache with the per-para chunk index vector.
+	/// Should only be called if `query_cache_for_para` returns `None`.
+	pub fn populate_for_para(
+		&mut self,
+		maybe_client_features: Option<ClientFeatures>,
+		babe_randomness: BabeRandomness,
+		n_validators: usize,
+		block_number: BlockNumber,
+		session_index: SessionIndex,
+		para_id: ParaId,
+	) -> Vec<ChunkIndex> {
+		let shuffle = Self::get_shuffle(
+			maybe_client_features.as_ref(),
+			block_number,
+			babe_randomness,
+			n_validators,
+		);
+		self.0.insert((block_number, session_index), (shuffle, maybe_client_features));
+
+		self.query_cache_for_para(block_number, session_index, para_id)
+			.expect("We just inserted the entry.")
+	}
+
+	fn get_shuffle(
+		maybe_client_features: Option<&ClientFeatures>,
+		block_number: BlockNumber,
+		mut babe_randomness: BabeRandomness,
+		n_validators: usize,
+	) -> Vec<ChunkIndex> {
+		let mut indices: Vec<_> = (0..n_validators)
+			.map(|i| ChunkIndex(u32::try_from(i).expect("validator count should not exceed u32")))
+			.collect();
+
+		if let Some(features) = maybe_client_features {
+			if features.contains(ClientFeatures::AVAILABILITY_CHUNK_SHUFFLING) {
+				let block_number_bytes = block_number.to_be_bytes();
+				for i in 0..32 {
+					babe_randomness[i] ^= block_number_bytes[i % block_number_bytes.len()];
+				}
+
+				let mut rng: ChaCha8Rng = SeedableRng::from_seed(babe_randomness);
+
+				indices.shuffle(&mut rng);
+			}
+		}
+
+		indices
+	}
+
+	/// Return the availability chunk start index for this para.
+	fn para_start_index(
+		maybe_client_features: Option<&ClientFeatures>,
+		n_validators: usize,
+		para_id: ParaId,
+	) -> usize {
+		if let Some(features) = maybe_client_features {
+			if features.contains(ClientFeatures::AVAILABILITY_CHUNK_SHUFFLING) {
+				let mut rng: ChaCha8Rng =
+					SeedableRng::from_seed(
+						u32::from(para_id).to_be_bytes().repeat(8).try_into().expect(
+							"vector of 32 bytes is safe to cast to array of 32 bytes. qed.",
+						),
+					);
+				return rng.gen_range(0..n_validators)
+			}
+		}
+
+		0
+	}
+
+	fn chunk_index_for_validator(
+		maybe_client_features: Option<&ClientFeatures>,
+		shuffle: &Vec<ChunkIndex>,
+		para_id: ParaId,
+		validator_index: ValidatorIndex,
+	) -> ChunkIndex {
+		let core_start_index =
+			Self::para_start_index(maybe_client_features, shuffle.len(), para_id);
+
+		let chunk_index = shuffle[(core_start_index +
+			usize::try_from(validator_index.0)
```

Comment:
nit: then why not validator_index.0 as usize.

Reply:
I avoid doing `as` casts simply because there is no error/panic in the (currently impossible) case where there is a failed conversion